packages/react/src/components/cookieConsent/modal/Modal.tsx

```diff
@@ -0,0 +1,74 @@
+import React, { useContext, useEffect, useState } from 'react';
+import ReactDOM from 'react-dom';
+import { VisuallyHidden } from '@react-aria/visually-hidden';
+
+import classNames from '../../../utils/classNames';
+import styles from '../CookieConsent.module.scss';
+import { CookieConsentContext, useCookieConsentUiTexts, useFocusShift } from '../CookieConsentContext';
+import { Content } from '../content/Content';
+import { useEscKey } from '../useEscKey';
+
+export function Modal(): React.ReactElement | null {
+  const cookieConsentContext = useContext(CookieConsentContext);
+  const hasOnlyRequiredConsents =
+    !cookieConsentContext.content.optionalCookies || cookieConsentContext.content.optionalCookies.groups.length === 0;
+  const shouldShowModal = !hasOnlyRequiredConsents && !cookieConsentContext.hasUserHandledAllConsents();
+  const [isModalInitiallyShown] = useState<boolean>(shouldShowModal);
+  const [popupTimerComplete, setPopupTimerComplete] = useState<boolean>(false);
+  const popupDelayInMs = 500;
+  // if hasUserHandledAllConsents() was false at first and then later true, user must have saved consents.
+  const showScreenReaderSaveNotification = isModalInitiallyShown && !shouldShowModal;
+  const { settingsSaved } = useCookieConsentUiTexts();
+  const containerId = 'HdsCookieConsentContainer';
+  const getContainerElement = (): HTMLElement | null => document.getElementById(containerId);
+  const [isDomReady, setIsDomReady] = useState<boolean>(false);
+
+  useEffect(() => {
+    setTimeout(() => setPopupTimerComplete(true), popupDelayInMs);
+  }, []);
+
+  useEffect(() => {
+    const containerElement = getContainerElement();
+    if (shouldShowModal && !isDomReady) {
+      if (!containerElement) {
+        const htmlContainer = document.createElement('div');
+        htmlContainer.setAttribute('id', containerId);
+        htmlContainer.setAttribute('data-testid', 'html-cookie-consent-container');
+        document.body.insertBefore(htmlContainer, document.body.firstChild);
+      }
+      setIsDomReady(true);
+    } else if (!shouldShowModal && isDomReady) {
+      if (containerElement) {
+        document.body.removeChild(containerElement);
```

Comment:
Just noticed that destroying an element directly from DOM does not call unmount listeners created in React.components. Meaning functions returned from useEffect() won't be called.

Seems like Portal should use useLayoutEffect() to make components work properly.

Would be nice to make a parent component which is the portal and it adds the modal.

Reply:
Very good catch! This needs to be fixed somehow. Now if you close the modal, and press esc, the heading element outside the modal will still be focused.

Reply:
> Just noticed that destroying an element directly from DOM does not call unmount listeners created in React.components. Meaning functions returned from useEffect() won't be called.

This is true, but we are lucky, because when the modal is closed, this Modal component will render the screen notification before the element is removed from DOM. Which means the removal of event listeners from useEffect  inside the `Content` component will be called.

But there was a real bug here, because this `Modal` component was never unmounted in our examples. Which led to a situation where the event listeners set in Modal's own useEffect were never removed. This is now fixed with one if-clause

Reply:
document.body.removeChild will throw an error when there is an existing container element which is not child of document.body. Fixed in Portal PR.

Reply:
The Portal PR also removes all keyup listeners